### DISCUSSION TITLE
Cluster expansion with zero down time

### DISCUSSION THREAD

**Discussion by sdukshis** [minio/minio#16748] [Q&A] [answered]
I'd like to add expand a cluster by adding another server pool as it's describe in documentation - https://min.io/docs/minio/linux/operations/install-deploy-manage/expand-minio-deployment.html?ref=docs-redirect#id4 

One of steps is to reboot all nodes simultaneously.

In my experiments while I've rebooting all nodes with a new config and warp benchmark running in parallel I've sometimes got a connection error.

My question is following: is it possible to expand cluster by adding a new server pool with zero downtime? If so how can I do this? 

For example. I've considering following configuration
Pool 1: http://minio{1..3}.example.net/mnt/disk1/minio
Pool 2: http://minio{4..6}.example.net/mnt/disk1/minio

HAProxy on separate node balancing on minio{1..3}.example.net

After /etc/default/minio config update. Can I restart only minio{4..6}, and after restart minio1, and after restart minio{2,3}. So in every time moment at least 1 minio instance will be working and serving requests from haproxy.

**Comment by klauspost** [comment, marked as answer]
The original "to be expanded pool" can be kept running while you bring up the "expansion pool". After that you might as well restart all nodes in the original pool at once.

>  I've sometimes got a connection error

If you run through the load balancer you shouldn't get connection errors, but some temporarily failed request.

S3 API clients should expect that occasional errors pop up and have appropriate retry/backoff mechanics. Even AWS S3 has occasional failed requests.

As such minor interruptions like restarts shouldn't affect operation of clients.

**Reply by danthegoodman1** [reply]
will the expansion pool be able to route requests to the original pool while the original pool has yet to be restarted with the new configuration?

For example I could configure the load balancer to only point to the new expansion pool, which effectively gives access to both pools, since any requests going to the original pool would be unable to find objects on the expansion pool until it is restarted.

This would be to minimize any number of read requests going to the original to objects that have written to the expansion pool before the original pool has restarted.

**Reply by klauspost** [reply]
@danthegoodman1 No.

**Reply by danthegoodman1** [reply]
So, if I understand correctly, there has to be some amount of downtime when expanding?

**Reply by klauspost** [reply]
Correct. You shouldn't plan expansions as an everyday event for this and other reasons.
